specs/union.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the humbug/php-scoper package.
+ *
+ * Copyright (c) 2017 Théo FIDRY <theo.fidry@gmail.com>,
+ *                    Pádraic Brady <padraic.brady@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+return [
+    'meta' => [
+        'title' => 'Union types',
+        // Default values. If not specified will be the one used
+        'prefix' => 'Humbug',
+
+        'expose-global-constants' => false,
+        'expose-global-classes' => false,
+        'expose-global-functions' => false,
+        'expose-namespaces' => [],
+        'expose-constants' => [],
+        'expose-classes' => [],
+        'expose-functions' => [],
+
+        'exclude-namespaces' => [],
+        'exclude-constants' => [],
+        'exclude-classes' => [],
+        'exclude-functions' => [],
+
+        'expected-recorded-classes' => [],
+        'expected-recorded-functions' => [],
+    ],
+
+    'Method casts' => <<<'PHP'
+    <?php
+    
+    class X
+    {
+        public function foo(Y|Z $a, null|Y $b) : null|Y|Z
+        {
+        }
+    }
+    
+    ----
+    <?php
+    
+    namespace Humbug;
+    
+    class X
+    {
+        public function foo(Y|Z $a, null|Y $b) : null|Y|Z
```

Comment:
might be worth adding `static` or `self` into the mix as well? Isn't there a nullable variant too?

If this is too bloated for a single method it's also find to have those cases within different functions:

```
        public function method0(Y|Z $a) : Y|Z
        public function method1(Y|null $a) : Y|null
// ...
```

Reply:
Added these cases as well and covered traits and interfaces as well.